src/dex/nerve/nerve-math.ts

```diff
@@ -0,0 +1,317 @@
+    else v.d0 = d0;
+
+    // v.d1 = v.d0.sub(tokenAmount.mul(v.d0).div(self.lpToken.totalSupply()));
+    v.d1 = v.d0 - (tokenAmount * v.d0) / state.lpToken_supply;
+    const newY = this._getYD(state, v.preciseA, tokenIndex, xp, v.d1);
+    if (newY === null) return null;
+    else v.newY = newY;
+
+    const xpReduced: bigint[] = [];
+
+    v.feePerToken = this._feePerToken(state);
+    for (let i = 0; i < numTokens; i++) {
+      const xpi = xp[i];
+      // if i == tokenIndex, dxExpected = xp[i] * d1 / d0 - newY
+      // else dxExpected = xp[i] - (xp[i] * d1 / d0)
+      // xpReduced[i] -= dxExpected * fee / FEE_DENOMINATOR
+
+      const dxExpected =
+        i === tokenIndex
+          ? (xpi * v.d1) / v.d0 - v.newY
+          : xpi - (xpi * v.d1) / v.d0;
+      xpReduced[i] -= (dxExpected * v.feePerToken) / this.FEE_DENOMINATOR;
+    }
+
+    const yd = this._getYD(state, v.preciseA, tokenIndex, xpReduced, v.d1);
+    if (yd === null) return null;
+    // uint256 dy = xpReduced[tokenIndex].sub(getYD(v.preciseA, tokenIndex, xpReduced, v.d1));
+    let dy = xpReduced[tokenIndex] - yd;
+
+    // dy = dy.sub(1).div(self.tokenPrecisionMultipliers[tokenIndex]);
+    dy = (dy - ONE) / state.tokenPrecisionMultipliers[tokenIndex];
+
+    return { dy, newY: v.newY };
+  }
+
+  protected _feePerToken(state: PoolState) {
+    const numTokens = this._getNumTokens(state);
+    // self.swapFee.mul(self.pooledTokens.length).div(self.pooledTokens.length.sub(1).mul(4));
+    return (state.swapFee * numTokens) / ((numTokens - ONE) * biginterify(4));
+  }
+
+  protected _calculateCurrentWithdrawFee(state: PoolState) {
+    // It is not correct. We should calculate user withdrawFeeMultiplier by
+    // the time passed since the liquidity was added
+    return state.defaultWithdrawFee ? state.defaultWithdrawFee : ZERO;
+  }
+
+  protected _getYD(
+    state: PoolState,
+    a: bigint,
+    tokenIndex: number,
+    xp: bigint[],
+    d: bigint,
+  ) {
+    const numTokens = biginterify(xp.length);
+
+    let c = d;
+    let s: bigint = ZERO;
+    const nA = a * numTokens;
+
+    for (let i = 0; i < numTokens; i++) {
+      if (i != tokenIndex) {
+        // s = s.add(xp[i]);
+        s = s + xp[i];
+
+        // c = c.mul(d).div(xp[i].mul(numTokens));
+        c = (c * d) / (xp[i] * numTokens);
+      }
+    }
+    // c = c.mul(d).mul(A_PRECISION).div(nA.mul(numTokens));
+    c = (c * d * this.A_PRECISION) / (nA * numTokens);
+
+    // uint256 b = s.add(d.mul(A_PRECISION).div(nA));
+    const b = s + (d * this.A_PRECISION) / nA;
+    let yPrev: bigint;
+    let y = d;
+
+    for (let i = 0; i < this.MAX_LOOP_LIMIT; i++) {
+      yPrev = y;
+
+      // y = y.mul(y).add(c).div(y.mul(2).add(b).sub(d));
+      y = (y * y + c) / (y * biginterify(2) + b - d);
+      if (MathUtil.within1(y, yPrev)) {
+        return y;
+      }
+    }
+
+    this.logger.error(`Event pool ${this.name} method _getYD did not converge`);
+    state.isValid = false;
+    return null;
```

Comment:
may be its better to throw error instead of logging and returning null. This will make the implementation more clean and also similar to the contract. You can have a try catch in the outer level to handle it

Reply:
Thanks, it looks much better. I didn't like this `null` checks. And I am glad to get rid of them.